packages/frontend/src/connectors/frame/index.ts

```diff
@@ -0,0 +1,124 @@
+        super(config);
+    }
+
+    public async connect({ chainId }: { chainId?: number } = {}): Promise<
+        Required<ConnectorData>
+    > {
+        let targetChainId = chainId;
+        let useDefaultChainId = false;
+        if (!targetChainId || this.isChainUnsupported(targetChainId)) {
+            try {
+                const lastUsedChainId = await this.getChainId();
+                if (
+                    lastUsedChainId &&
+                    !this.isChainUnsupported(lastUsedChainId)
+                ) {
+                    targetChainId = lastUsedChainId;
+                } else {
+                    useDefaultChainId = true;
+                }
+            } catch (e) {
+                useDefaultChainId = true;
+            }
+        }
+
+        const provider = await this.getProvider({
+            chainId: targetChainId,
+            create: true,
+        });
+        provider.on("accountsChanged", this.onAccountsChanged);
+        provider.on("chainChanged", this.onChainChanged);
+        provider.on("disconnect", this.onDisconnect);
+
+        const id = useDefaultChainId
+            ? this.chains[0].id
+            : await this.getChainId();
+
+        const unsupported = this.isChainUnsupported(id);
+
+        const signer = await provider.getSigner();
+        return {
+            account: (await signer.getAddress()) as Address,
+            chain: { id, unsupported },
+            provider,
+        };
+    }
+
+    public async disconnect() {
+        const provider = await this.getProvider();
+
+        provider.removeListener("accountsChanged", this.onAccountsChanged);
+        provider.removeListener("chainChanged", this.onChainChanged);
+        provider.removeListener("disconnect", this.onDisconnect);
+    }
+
+    public async getAccount(): Promise<Address> {
+        const provider = await this.getProvider();
+        const signer = await provider.getSigner();
+        return (await signer.getAddress()) as Address;
+    }
+
+    public async getChainId() {
+        const provider = await this.getProvider();
+        const network = await provider.getNetwork();
+        const chainId = normalizeChainId(network.chainId);
+        return chainId;
+    }
+
+    public async getProvider({
+        chainId,
+        create,
+    }: { chainId?: number; create?: boolean } = {}) {
+        if (!this.provider || chainId || create) {
+            this.provider = new JsonRpcProvider(
+                "http://127.0.0.1:1248",
+                chainId
+            );
+        }
+        return this.provider;
+    }
+
+    async getSigner() {
+        const provider = await this.getProvider();
+        return provider.getSigner();
+    }
+
+    async isAuthorized() {
+        return true;
+    }
+
+    async switchChain(chainId: number): Promise<Chain> {
+        await this.getProvider({ chainId, create: true });
+        this.onChainChanged(chainId);
```

Comment:
Should we move `this.onChainChanged(chainId);` below the if `(!chain) ...`?

Reply:
U rite, and I've also fixed the readonly connector in1a52d868580bfc72e64f48597a2fbec0dd25640b